ws-nextjs-app/pages/[service]/send/[id]/FormContext/index.tsx

```diff
@@ -0,0 +1,104 @@
+import React, {
+  createContext,
+  FormEvent,
+  PropsWithChildren,
+  useContext,
+  useState,
+} from 'react';
+import { FetchError } from '#app/models/types/fetch';
+import { v4 as uuid } from 'uuid';
+
+import { useRouter } from 'next/router';
+import { OK } from '#app/lib/statusCodes.const';
+import {
+  Field,
+  OnChangeHandler,
+  OnChangeInputName,
+  OnChangeInputValue,
+} from '../types';
+
+type SubmissionError = {
+  message: string;
+  status: number;
+} | null;
+
+type ContextProps = {
+  formState: Record<OnChangeInputName, OnChangeInputValue | null>;
+  handleChange: OnChangeHandler;
+  handleSubmit: (event: FormEvent) => Promise<void>;
+  submissionError?: SubmissionError;
+};
+
+const FormContext = createContext({} as ContextProps);
+
+const getInitialFormState = (
+  fields: Field[],
+): Record<OnChangeInputName, OnChangeInputValue | null> =>
+  fields?.reduce((acc, field) => ({ ...acc, [field.id]: null }), {});
+
+export const FormContextProvider = ({
+  fields,
+  children,
+}: PropsWithChildren<{ fields: Field[] }>) => {
+  const {
+    query: { id },
+  } = useRouter();
+
+  const [formState, setFormState] = useState(getInitialFormState(fields));
+  const [submissionError, setSubmissionError] = useState<SubmissionError>(null);
+
+  const handleChange = (name: OnChangeInputName, value: OnChangeInputValue) => {
+    setFormState(prevState => {
+      return { ...prevState, [name]: value };
+    });
+  };
+
+  const handleSubmit = async (event: FormEvent) => {
+    event.preventDefault();
+
+    // Reset error state
+    setSubmissionError(null);
+
+    const formData = new FormData();
+
+    // TODO: This is a mock data, we should use the formState instead
+    Object.entries({ surname: 'BBC TEST NAME' }).forEach(([key, value]) => {
+      formData.append(key, value);
+    });
+
+    try {
+      const url = `https://www.bbc.com/ugc/send/${id}?said=${uuid()}`;
+
+      const req = new XMLHttpRequest();
```

Comment:
Are we using XMLHttpRequest API here for browser compatibility reasons over the fetch API?

Reply:
Good question. Its mainly to support the `onprogress` event that we'll need for file uploads: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/progress_event. The native fetch doesn't expose event handlers for `onprogress`, whereas the older XMLHttpRequest does.